Code/DataStructs/MultiFPBReader.cpp

```diff
@@ -0,0 +1,293 @@
+        return v1.get<2>() < v2.get<2>();
+      }
+    } else {
+      return v1.get<0>() > v2.get<0>();
+    }
+  }
+};
+struct pairSorter
+    : public std::binary_function<std::pair<unsigned int, unsigned int>,
+                                  std::pair<unsigned int, unsigned int>, bool> {
+  bool operator()(const std::pair<unsigned int, unsigned int> &v1,
+                  const std::pair<unsigned int, unsigned int> &v2) const {
+    if (v1.first == v2.first) {
+      return v1.second < v2.second;
+    } else {
+      return v1.first < v2.first;
+    }
+  }
+};
+
+struct sim_args {
+  const boost::uint8_t *bv;
+  double ca, cb;
+  double threshold;
+  const std::vector<FPBReader *> &readers;
+  std::vector<std::vector<MultiFPBReader::ResultTuple> > *res;
+  bool initOnSearch;
+};
+
+void tversky_helper(unsigned int threadId, unsigned int numThreads,
+                    const sim_args *args) {
+  for (unsigned int i = threadId; i < args->readers.size(); i += numThreads) {
+    if (args->initOnSearch) args->readers[i]->init();
+    std::vector<std::pair<double, unsigned int> > r_res =
+        args->readers[i]->getTverskyNeighbors(args->bv, args->ca, args->cb,
+                                              args->threshold);
+    (*args->res)[i].clear();
+    (*args->res)[i].reserve(r_res.size());
+    for (std::vector<std::pair<double, unsigned int> >::const_iterator rit =
+             r_res.begin();
+         rit != r_res.end(); ++rit) {
+      (*args->res)[i].push_back(
+          MultiFPBReader::ResultTuple(rit->first, rit->second, i));
+    }
+  }
+}
+void tani_helper(unsigned int threadId, unsigned int numThreads,
+                 const sim_args *args) {
+  for (unsigned int i = threadId; i < args->readers.size(); i += numThreads) {
+    if (args->initOnSearch) args->readers[i]->init();
+    std::vector<std::pair<double, unsigned int> > r_res =
+        args->readers[i]->getTanimotoNeighbors(args->bv, args->threshold);
+    (*args->res)[i].clear();
+    (*args->res)[i].reserve(r_res.size());
+    for (std::vector<std::pair<double, unsigned int> >::const_iterator rit =
+             r_res.begin();
+         rit != r_res.end(); ++rit) {
+      (*args->res)[i].push_back(
+          MultiFPBReader::ResultTuple(rit->first, rit->second, i));
+    }
+  }
+}
+
+template <typename T>
+void generic_nbr_helper(std::vector<MultiFPBReader::ResultTuple> &res, T func,
+                        const sim_args &args, unsigned int numThreads) {
+  res.clear();
+  res.resize(0);
+  numThreads = getNumThreadsToUse(numThreads);
+#ifdef RDK_THREADSAFE_SSS
+  boost::thread_group tg;
+#endif
+  if (numThreads == 1) {
+    func(0, 1, &args);
+  }
+#ifdef RDK_THREADSAFE_SSS
+  else {
+    for (unsigned int tid = 0; tid < numThreads && tid < args.readers.size();
+         ++tid) {
+      tg.add_thread(new boost::thread(func, tid, numThreads, &args));
+    }
+    tg.join_all();
+  }
+#endif
+
+  for (unsigned int i = 0; i < args.readers.size(); ++i) {
```

Comment:
Yes.  If these are small it won't matter though.

----
Brian Kelley

> On May 20, 2016, at 7:48 AM, Greg Landrum <notifications@github.com> wrote:
> 
> In Code/DataStructs/MultiFPBReader.cpp:
> 
> > +  boost::thread_group tg;
> > +#endif
> > +  if (numThreads == 1) {
> > +    func(0, 1, &args);
> > +  }
> > +#ifdef RDK_THREADSAFE_SSS
> > +  else {
> > +    for (unsigned int tid = 0; tid < numThreads && tid < args.readers.size();
> > +         ++tid) {
> > +      tg.add_thread(new boost::thread(func, tid, numThreads, &args));
> > +    }
> > +    tg.join_all();
> > +  }
> > +#endif
> > +
> > +  for (unsigned int i = 0; i < args.readers.size(); ++i) {
> You mean in order to save calls to reserve()?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
>